Fleshed out communication layer between embeddables and dashboard

stacey-gammon opened the pull request:
Introduce an improved communication flow for embeddables and dashboard. This sets the stage for things like per panel time range, filters and drilldown links.

View the README.md for an overview.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1203/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1224/)

Stacey-Gammon commented on src/core_plugins/kibana/public/dashboard/__tests__/dashboard_state.js:
```diff
@@ -1,102 +0,0 @@
-import ngMock from 'ng_mock';
```
Migrated these to jest tests and modified the index pattern ones to handle the new communication of index pattern from embeddables.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1226/)

Stacey-Gammon commented on src/core_plugins/kibana/public/dashboard/dashboard_app.js:
```diff
@@ -212,19 +210,16 @@ app.directive('dashboardApp', function ($injector) {
 
-      $scope.$listen(timefilter, 'fetch', $scope.refresh);
+      $scope.$listen(timefilter, 'fetch', () => {
+        dashboardStateManager.handleTimeChange($scope.timefilter);
```
We can't yet get rid of scope.refresh and the `$rootScope.$broadcast('fetch');` call because changing filters/query still goes through there.  Once we also shuffle that logic through to the embeddables directly (as desired for https://github.com/elastic/kibana/issues/16641) we should be able to get rid of this crazy, fragile async logic.

Stacey-Gammon replied:
Ah I was wrong about this, discover/embedded search needs the broadcast fetch call because it's not listening directly to the time filter passed down by the dashboard - but instead to the global one.  I added a test to catch the bug caused by removing this.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1229/)

Stacey-Gammon commented on src/core_plugins/kibana/public/dashboard/dashboard_container_api.js:
```diff
@@ -1,26 +0,0 @@
-import { ContainerAPI } from 'ui/embeddable';
```
Instead of communicating via this API and using a slew of functions, this has been replaced via communication with state only, and two functions, one for each communication direction.

Stacey-Gammon commented on src/core_plugins/kibana/public/dashboard/dashboard_state_manager.js:
```diff
@@ -128,27 +126,8 @@ export class DashboardStateManager {
-   * ... until then... we are stuck with this manual crap. :(
-   * Fixes https://github.com/elastic/kibana/issues/15720
-   */
-  triggerEmbeddableConfigUpdateListeners() {
```
These changes now get sent to the embeddable via an onContainerStateChanged call when the EmbeddableViewport detects a change in the return value of getContainerState.

Stacey-Gammon commented on src/core_plugins/kibana/public/dashboard/dashboard_state_manager.js:
```diff
@@ -210,12 +187,29 @@ export class DashboardStateManager {
     if (!this._areStoreAndAppStatePanelsEqual()) {
       const panels = getPanels(store.getState());
       this.appState.panels = [];
+      this.panelIndexPatternMapping = {};
```
Rather than do anything fancy with detecting added vs removed vs changed panels, we're just clearing and rebuilding below

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1254/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1264/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1266/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1270/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1271/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1306/)

Stacey-Gammon commented on src/core_plugins/kibana/public/dashboard/actions/embeddables.js:
```diff
@@ -1,58 +1,74 @@
-  return (dispatch) => {
+export function initializeEmbeddable({ embeddableFactory, panelId }) {
+  return (dispatch, getState) => {
+    const panel = getPanel(getState(), panelId);
```
So, I think the one downside to your proposed approach is that we'd be moving the async call into the react component, which I've been trying to get away from because some deem it an anti pattern.  If you look at this: https://codebrahma.com/structuring-async-operations-react-redux-applications/ , this persons `How we can improve from here ?` part where they say to use redux and middleware for the async calls seems to be the pattern I'm following here.

If it weren't for the functions stored in the cache, I could put the rest of the metadata in redux and don't have a problem with that.

Is there an approach I'm missing where I can keep the async fetches in the actions, but still get rid of the cache here?

Stacey-Gammon replied:
> n DashboardPanel's componentDidMount (maybe), we'd create the underling embeddable and insert it into the DOM, passing it the initialization data that was retrieved by connect's mapStateToProps, and passing it a state change handler that would look something like (state) => this.onEmbeddableStateChange(state) or something

It's this part that I think I disagree with.  I like that initialize returns an Embeddable instance with the render, destroy and onPanelStateChanged functions.  I like that I can call onPanelStateChanged without needing the embeddable to even be rendered (so it's kept up to date for when it eventually is rendered on screen).  I also like that I don't have to pass the embeddableFactory along.  I think what you are proposing is having `embeddableFactory.initialize(id, data) => metadata` then later calling `embeddableFactory.render(id, data, onContainerStateChanged) => onEmbeddableStateChanged`  Having two calls on the embeddableFactory feels odd to me.  I like the single create call which returns an Embeddable instance for all our needs.

I could remove the render method on the embeddableHandlerCache and just let anyone grab the Embeddable object, so that all actual render and destroy calls are inside the react component, though anyone could still grab the embeddable object and call render/destroy wherever they wanted.  But maybe that wouldn't feel as wrong?  So instead of
```
  async componentDidMount() {
    embeddableHandlerCache.render(this.props.panelId, this.panelElement);
  }
```

it'd be
```
  async componentDidMount() {
    embeddableHandlerCache.get(this.props.panelId).render(this.panelElement);
  }
```

Essentially the same thing but at least all references to "render" (and we could do the same with "destroy" are in the component?

Stacey-Gammon commented on src/core_plugins/kibana/public/dashboard/actions/embeddables.js:
```diff
@@ -1,58 +1,74 @@
-      .then(embeddable => {
-        return dispatch(embeddableRenderFinished(panel.panelIndex, embeddable));
+    dispatch(embeddableIsInitializing(panelId));
+    embeddableFactory.create(panel, embeddableStateChanged.bind(this, dispatch, getState, panelId))
```
Nice, i like it!

Stacey-Gammon commented on src/core_plugins/kibana/public/dashboard/actions/embeddables.js:
```diff
@@ -1,58 +1,74 @@
+  }
+
+  // TODO: communicate time range and filter changes from the embeddable through to dashboard rather than
+  // through angular components.
```
I meant to leave it in but I can get rid of.  I think technically you are right and leaving this in is anti our style guide.

Stacey-Gammon commented on src/core_plugins/kibana/public/dashboard/cache/embeddable_handler_cache.js:
```diff
@@ -0,0 +1,55 @@
+   * @param {Element} domNode
+   */
+  render(panelId, domNode) {
+    this.cache[panelId].render(domNode);
```
yea, except that I can't seem to get around it while keeping the async initialization part out of react and in the actions. If I move the initialization portion into the component then I'm coupling them and can't initialize prior to rendering, perhaps to speed up/cache some stuff - e.g. if there are multiple tabs, I could initialize all embeddables, but not render the ones on the second page.

Stacey-Gammon commented on src/core_plugins/kibana/public/dashboard/panel/dashboard_panel_container.js:
```diff
@@ -5,39 +5,34 @@ import { DashboardPanel } from './dashboard_panel';
 import { DashboardViewMode } from '../dashboard_view_mode';
```
Why we are wrapping dashboard_panel in a container?  To map the state and props from redux.  Unless you don't think it's necessary? Or maybe I'm misunderstanding your question.

Stacey-Gammon commented on src/core_plugins/kibana/public/dashboard/reducers/embeddables.js:
```diff
@@ -1,13 +1,56 @@
-  [destroyEmbeddable]:
+  [clearStagedFilters]:
+    (embeddables) => {
+      return _.reduce(embeddables, (embeddablesCopy, embeddable, key) => {
```
Ah, nice.  I think I might still need to do { ...embeddable } to ensure a copy is made...

Stacey-Gammon commented on src/core_plugins/kibana/public/dashboard/reducers/embeddables.js:
```diff
@@ -1,13 +1,56 @@
+  /**
+   *
+   * @param embeddables {Object.<string, EmbeddableState>}
+   * @param payload {String} The id of the panel whose embeddable should be loaded
```
Oh I dislike this too.  I think it has to be named payload.  Or I could do payload.embeddableId.  But I think payload is automatically the name when you use `createAction` from `redux-actions`?  Whether or not we should be using redux-actions is another question....  I'm not sure it helps more than it obscures.  But not really a decision to be made for this PR.

Stacey-Gammon commented on src/core_plugins/kibana/public/dashboard/selectors/index.js:
```diff
@@ -38,10 +41,14 @@ export const getPanel = (dashboard, panelId) => getPanels(dashboard)[panelId];
 
+export const getEmbeddables = (dashboard) => dashboard.embeddables;
+
+export const getEmbeddablePersonalization = (dashboard, panelId) => getPanel(dashboard, panelId).embeddableConfig;
```
It's a per dashboard panel setting, not per user.  I like the name `personalization` but for BWC can't rename the other thing from `embeddableConfig` without some migration work.

I can batch rename embeddablePersonalization to embeddableConfig... or live with the mixed naming for now.  What would you prefer?  I'm pretty ambivalent on either option.

Stacey-Gammon commented on src/core_plugins/kibana/public/discover/embeddable/search_embeddable.js:
```diff
@@ -0,0 +1,122 @@
+import searchTemplate from './search_template.html';
+import * as columnActions from 'ui/doc_table/actions/columns';
+
+export class SearchEmbeddable extends Embeddable {
```
You mean like make the entire search embeddable be a react/redux component instead of angular? Not without a lot of work.  Unless you are talking about something much smaller?

Stacey-Gammon commented on test/functional/apps/dashboard/_bwc_shared_urls.js:
```diff
@@ -38,7 +40,7 @@ export default function ({ getService, getPageObjects }) {
       it('loads an unsaved dashboard', async function () {
         const url = `${kibanaBaseUrl}#/dashboard?${urlQuery}`;
-
+        log.debug(`Navigating to ${url}`);
```
I did.  It's a lot easier to debug when this fails and I have the url handy, I can navigate to it directly. I know I can get the url from above, but with all the whitespace stuff I can't easily copy/paste it into the browser.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1366/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1445/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1455/)

Stacey-Gammon commented:
Okay, @chrisdavies ended up talking me into getting rid of the embeddable cache and putting all render/destroy handling only inside the react component. To reiterate, reasons are:
 - less chance of accidentally double destroying
 - UI functions like render and destroy only available in react
 - lets redux handle all metadata and updating
 - my argument for keeping async actions out of components and into redux is more a recommendation for data, not ui element stuff and pros outweigh cons.
 - my argument for being able to pre-cache or make smarter performance decisions is premature. We have no concrete ideas or plans to do something like this right now and it's all theoretically.  Best not to use an implementation on a vague futuristic idea.

So, this way wins!  At least for the time being.  :D  And benefit is the embeddable/container API stays the same, only where it resides in the container has changed.

Good for another look!

oh and I found a bug while I was updating this, so I added a test that would have caught it while i fixed it.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1474/)

Stacey-Gammon commented:
cc @uboness - lmk if you have any questions.  I have a README.md in here that starts to explain the state and communication and some pseudo code for how it's used.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1597/)

Stacey-Gammon commented:
Had a couple conversations recently with @uboness and @timroes which made me rethink some of this structure.  Latest thoughts:

## State vs actions
The current approach lands pretty heavily on using pure state and no "actions", or function calls.  I have doubts whether this approach will last the long haul.  Consider the way stagedFilters is currently implemented via a state change.  This feels somewhat awkward and forced:

 1. Embeddable sets `stagedFilters` in embeddableState
 2. Container is notified of the change, see's stagedFilters has contents in the embeddable state, applies the filters, clears the stagedFilters state.  

stagedFilters state **does not persist the next time the embeddable is updated**.  The action based way of doing this would be to call a `container.applyFilters(filters)` which feels more straightforward than awkwardly passing a parameter and telling a function to be called via a state tree.

Consider another scenario - a way to tell the embeddable to draw itself on an HTML canvas element.  How could we accomplish this via pure state?  Set a flag `drawOnCanvas` to true in containerState, which gets sent to the embeddable, which must detect that is set to `true`, which must then return this information in the form of state to dashboard?  You might be able to break this down into pure state, but it's very awkward, especially the two way communication flow via state trees.  Instead, a more straightforward approach would be to add a new function on the Embeddable - `draw`.  Perhaps include in the state a new parameter `isDrawable: boolean` to tell dashboard that this embeddable has that functionality.

Yet another scenario: create a machine learning job.  What if the information machine learning needs when this action is initiated is expensive, and doesn't need to persist? It then seems like something we expose on the embeddable in the form of a function instead of passing via state (e.g. `embeddable.generateEsQuery()` instead of `embeddableState.esQuery`)

So my thoughts on state vs actions is:
- If it's expensive or transitory, communicate via action functions.
- If it persists, and is something you could see being in a url, communicate via state

## Static vs Dynamic state (state vs metadata)
I no longer think the separation of embeddableMetadata and embeddableState is necessary.  We should assume and support the case of any state changing, including title, editUrl and indexPatterns.  Eventually we'd like to expose in place editing of a visualization on the dashboard.  This means title could change, index patterns, etc.  What does the separation of metadata and state give us expect to create an issue down the line when we realize one of our assumptions was wrong?

## Embeddable life cycle.
Iteration 1 of this PR de-coupled the rendering of the Embeddable from the react component - so the Embeddable could live outside of the react element.  Iteration 2 put them back together because this felt like it would be harder to introduce bugs (double destroying the embeddable).

I think I can still stick with this second iteration but while exploring pluggable panel actions, I realized I'll need to expose the embeddable to the panel header/menu options, so I'll have to create it farther up the chain (in DashboardPanel instead of EmbeddableViewport).

Stacey-Gammon commented:
Had another sync with @timroes and @chrisdavies this am regarding above thoughts.  As Chris summarized, the most important things are:

- Any components that we don't own (e.g. embeddables, visualizations, etc) cannot be tied to any framework-specific interface in any way
- We need to define a vanilla JS interface that such components are expected to implement
- We should be able to swap out our internal state management / rendering strategy at any time without impacting the way such components are written

Additionally, while we need to be flexible with a functional approach, for now, I'm going to leave stagedFilters specifically as state (this particular state is debatable - whether it should go in redux or be used as a function).

As such, I'm going to make the final two changes:
 - Pushing embeddable up to dashboard_panel so it can be passed down to the options menu for pluggable panel actions.
 - Combining embeddable metadata and state. 

Then this PR should be good to go for another look.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1626/)

Stacey-Gammon commented:
I hate to do this but I'm changing my mind again on splitting out static vs dynamic state.  Playing my own devils advocate:

> What does the separation of metadata and state give us expect to create an issue down the line when we realize one of our assumptions was wrong?

Forcing some state to be static throughout the lifecycle of an embeddable can enforce a stable and consistent UX experience.  Consider the soon(ish) to be introduced `supportsDrilldowns` flag.  If this is true, we'll show, in the panel context menu, some drill down link configuration options.  If this can change over the course of looking at a dashboard then you can get a weird experience where of all of a sudden a panel context menu option disappears.  It's less flexible to tell an embeddable "you can only give me this piece of data once", but it can create a better experience.  So I reverse my decision. I like the separation. We can debate which piece of data belongs where, and we can always move them around, but I think it makes sense for this differentiation to exist.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1639/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1643/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1673/)

nreese commented:
Error in console when creating a filter from a saved search. The error was created by clicking `filter for value` icon in the time column of the saved search

<img width="2047" alt="screen shot 2018-04-11 at 8 49 49 am" src="https://user-images.githubusercontent.com/373691/38624506-7169cf32-3d65-11e8-9e3b-b746020530f3.png">

Stacey-Gammon commented:
great catch @nreese!  I'll investigate but a test to catch it might be hard to add since from the ui it actually, eventually, does the right thing.  Just takes awhile to get there. And I don't think we monitor for console errors in our selenium tests...  anyway, thanks for the good find!

Stacey-Gammon commented:
ah i got it to kill the ui, i will add a test as well.

nreese commented:
When I update the global time range, the dashboard panels are not fetching new data

nreese commented:
Found another error, this time when the embeddable does not have an index pattern. To reproduce, create a dashboard with a vis that has an index pattern and a vis that does not have an index pattern (like markdown).

<img width="2035" alt="screen shot 2018-04-11 at 1 00 55 pm" src="https://user-images.githubusercontent.com/373691/38637514-92b140e4-3d88-11e8-8f0d-11b1562dc6c0.png">

nreese commented on src/core_plugins/kibana/public/dashboard/dashboard_state.test.js:
```diff
@@ -0,0 +1,110 @@
+    });
+
+    test('registers index pattern when an embeddable is initialized with one', async () => {
+      simulateNewEmbeddableWithIndexPattern('foo1', mockIndexPattern);
```
Maybe too picky, but it took me a few to figure out if the id `foo1` was the index Id or the panel Id. Maybe rename to `panelId1` for clarity.

Stacey-Gammon replied:
A good reason for passing multiple params via an object.  Hows this?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1731/)

Stacey-Gammon commented on src/core_plugins/kibana/public/visualize/embeddable/visualize_embeddable_factory.js:
```diff
@@ -1,101 +1,59 @@
+    const editUrl = this.getEditPath(visId);
+
+    const waitFor = [ getVisualizeLoader(), this.savedVisualizations.get(visId) ];
     return this.Promise.all(waitFor)
```
Great point. We should be using standard promises, though historically it's been angular promises everywhere.  As of  https://github.com/elastic/kibana/issues/13855 we want to use native.  Nathan actually has a WIP PR  that removes these, but I can take care of it here too (since this is all refactored anyway).

Stacey-Gammon commented on src/ui/public/embeddable/embeddable.js:
```diff
@@ -1,6 +1,40 @@
+
+    if (config.render) {
+      this.render = config.render;
+    }
```
Yea, they are all configurable but render was the only one I am currently passing in, I can add code for the others too. ++

Stacey-Gammon commented on src/ui/public/embeddable/embeddable_factory.js:
```diff
@@ -1,33 +1,30 @@
+   */
+  create(/* containerMetadata, onEmbeddableStateChanged*/) {
+    throw new Error('Must implement initialize.');
   }
```
Intended.  A way to indicate that if you are attempting to write an embeddable plugin and aren't overriding this function, you are doing it wrong.

Stacey-Gammon commented:
For some reason it's not letting me respond to one of your comment inline, so here it is (@chrisdavies):

> Personalization, to me, indicates this is a person-based setting. I think embeddableConfig or panelSetting or something would be clearer. But it's a minor point. Do what you think is best!

Yea I'm torn on the naming.  config, setting, state, data... they are all so similar.  This type of data is unique in that it is stored with the dashboard, per panel.  So I like personalization in that it's different than all the other synonyms we are using, like metadata vs state. containerCustomization?

Stacey-Gammon commented on src/core_plugins/kibana/public/dashboard/panel/dashboard_panel.js:
```diff
@@ -71,17 +133,28 @@ export class DashboardPanel extends React.Component {
+  destroy: PropTypes.func.isRequired,
+  containerState: PropTypes.shape({
+    timeRange: PropTypes.object.isRequired,
+    embeddablePersonalization: PropTypes.object.isRequired,
```
Removed it for embeddablePersonalization/cCustomization... timeRange I think should be required.  We don't set the default the timepicker does, so if we don't pass something in here, something is broken.  It actually was broken earlier when I wasn't setting timeRange on first loading, so I made it required here which would have caught that bug.  There are actually visualizations and index patterns that aren't time based, but I'm pretty sure in dashboard, there will always be a time picker.

Stacey-Gammon commented on src/core_plugins/kibana/public/visualize/embeddable/visualize_embeddable.js:
```diff
@@ -0,0 +1,115 @@
+    const parsedUiState = savedVisualization.uiStateJSON ? JSON.parse(savedVisualization.uiStateJSON) : {};
+    this.uiState = new PersistedState({
+      ...parsedUiState,
+    });
```
Hmm, yea I'm not sure if there wa s areason for this. I'll switch it and see how tests behave.  If there is a reason for it, I should have written a comment since it's not obvious.

nreese commented:
> Yea I'm torn on the naming. config, setting, state, data... they are all so similar. This type of data is unique in that it is stored with the dashboard, per panel. So I like personalization in that it's different than all the other synonyms we are using, like metadata vs state. containerCustomization?

I think the existing key of `embeddableConfig` is very clear.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1738/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1765/)

Stacey-Gammon commented:
Failed on 
```
fail: "context app context link in discover should open the context view with the selected document as anchor"
13:24:38      │        tryForTime timeout: Error: tryForTime timeout: [POST http://localhost:9515/session/8fec61533213454815c1625eda499159/element / {"using":"css selector","value":"[data-test-subj~=\"docTable\"]"}] no such element: Unable to locate element: {"method":"css selector","selector":"[data-test-subj~="docTable"]"}
13:24:38      │         (Session info: chrome=65.0.3325.162)
13:24:38      │         (Driver info: chromedriver=2.36.540471 (9c759b81a907e70363c6312294d30b6ccccc2752),platform=Linux 3.10.0-514.21.1.el7.x86_64 x86_64)
13:24:38      │           at Server._post (/var/lib/jenkins/workspace/elastic+kibana+pull-request+multijob-selenium/test/functional/services/rem
```
https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1765/

I suspect flakiness 

Jenkins, test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1766/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1767/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1792/)
